### PR TITLE
Starting work for better data syncing

### PR DESCRIPTION
**Pull request by thinkslynk**

### CODE REVIEW

File: src/main/kotlin/com/refinedmods/refinedstorage/data/sync/BiSyncedData.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    override fun unregisterClient() {
+        if(isClient) ClientSidePacketRegistry.INSTANCE.unregister(identifier)
+    }
+    override fun unregisterServer() {
+        if(!isClient) ServerSidePacketRegistry.INSTANCE.unregister(identifier)
+    }
+
+    override fun onUpdate() {
+        send()
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/data/sync/BiSyncedData.kt]: What is this for?

File: src/main/kotlin/com/refinedmods/refinedstorage/data/sync/C2SSyncedData.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+) : BiSyncedData<T>(identifier, isClient, internalData, serializer, player, onChanged) {
+
+    override fun send() {
+        val byteBuffer = byteBuffers.buffer()
+        try {
+            val buf = PacketByteBuf(byteBuffer)
+            this.serializer.write(buf, internalData)
+            when(isClient) {
+                true -> ClientSidePacketRegistry.INSTANCE.sendToServer(identifier, buf)
+                false -> {} // NO-OP on server
```
Comment (Kroppeb) [inline src/main/kotlin/com/refinedmods/refinedstorage/data/sync/C2SSyncedData.kt]: Maybe check this first, before the rest of the logic?

Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/data/sync/C2SSyncedData.kt]: It's inside the is Client check. In a C2S setup (client to server) only the client tracks changes and notifies the server. The server only receives updates and thus has a noop for their send.

File: src/main/kotlin/com/refinedmods/refinedstorage/data/sync/BiSyncedData.kt (hunk 1 above)
Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/data/sync/BiSyncedData.kt]: We track if the object itself is replaced, but if an internal field is updated we won't know. So I've set up an observable pattern where when an object's internals have changed, they notify their observer.

File: src/main/kotlin/com/refinedmods/refinedstorage/data/sync/C2SSyncedData.kt (hunk 2 above)
Comment (Kroppeb) [reply src/main/kotlin/com/refinedmods/refinedstorage/data/sync/C2SSyncedData.kt]: I mean convert
```kotlin
override fun send() {
        val byteBuffer = byteBuffers.buffer()
        try {
            val buf = PacketByteBuf(byteBuffer)
            this.serializer.write(buf, internalData)
            when(isClient) {
                true -> ClientSidePacketRegistry.INSTANCE.sendToServer(identifier, buf)
                false -> {} // NO-OP on server
            }
        } finally { byteBuffer.release() }
    }
```
into
```kotlin
    override fun send() {
        if(!isClient) return
        val byteBuffer = byteBuffers.buffer()
        try {
            val buf = PacketByteBuf(byteBuffer)
            this.serializer.write(buf, internalData)
            ClientSidePacketRegistry.INSTANCE.sendToServer(identifier, buf)
        } finally { byteBuffer.release() }
    }
```

Comment (thinkslynk) [reply src/main/kotlin/com/refinedmods/refinedstorage/data/sync/C2SSyncedData.kt]: Yeah, I was think about that while typing the first comment; no point in writing out to a buffer only to discard it. I agree.

**Pull request merged**